File: backend/services/investigationService.js
import db from "../config/db.js";
import * as investigationRepo from "../repositories/investigationRepository.js";
import * as locationRepo from "../repositories/locationRepository.js";
import * as stockRepo from "../repositories/stockMovementRepository.js";
import AppError from "../utils/AppError.js";

/**
 * Service to find duplicate stock transactions.
 *
 * @param {Object} filters
 * @param {string} [filters.startDate]
 * @param {string} [filters.endDate]
 * @param {string} [filters.includeNotes]
 * @param {string} [filters.excludeNotes]
 * @param {string} [filters.movementType]
 * @param {string} [filters.productName]
 * @param {string} [filters.username]
 * @param {string} [filters.location]
 * @returns {Promise<Array>}
 */
export const getDuplicateTransactionsService = async (filters, page = 1, limit = 10) => {
  const connection = await db.getConnection();
  try {
    // Validasi basic
    if (filters.startDate && filters.endDate) {
      if (new Date(filters.startDate) > new Date(filters.endDate)) {
        throw new AppError("Tanggal mulai tidak boleh lebih besar dari tanggal akhir.", 400);
      }
    }

    const offset = (page - 1) * limit;

    const totalGroups = await investigationRepo.countDuplicateGroups(connection, filters);
    const duplicates = await investigationRepo.getDuplicateGroups(
      connection,
      filters,
      limit,
      offset,
    );

    const invoiceSet = new Set();
    const itemIdSet = new Set();
    const invoiceRegex = /Sale Ref:\s+(.*?)\s+\(Item/i;

    // Grouping for better frontend consumption
    // We group by baseNote (Invoice Reference)
    const grouped = duplicates.reduce((acc, curr) => {
      const baseNote = curr.notes ? curr.notes.split(" (Item")[0].trim() : "Unknown";
      // Convert baseNote to uppercase to match MySQL case-insensitive grouping
      const normalizedBaseNote = baseNote.toUpperCase();
      const key = `${normalizedBaseNote}_${curr.movement_type}${
        filters.exactQuantity ? "_exact" : ""
      }`;

      if (!acc[key]) {
        let extractedInvoice = null;
        if (curr.notes) {
          const matchInv = curr.notes.match(invoiceRegex);
          if (matchInv && matchInv[1]) {
            extractedInvoice = matchInv[1].trim();
            invoiceSet.add(extractedInvoice);
          }
        }

        acc[key] = {
          baseNote: baseNote, // Keep original casing for display
          movementType: curr.movement_type,
          extractedInvoice,
          pickingList: null,
          totalQuantity: 0,
          totalTransaction: 0,
          uniqueItemsCount: 0,
          occurrences: 0, // Will be calculated based on unique timestamps
          transactions: [],
        };
      }

      acc[key].totalQuantity += curr.quantity;

      acc[key].transactions.push({
        id: curr.id,
        productId: curr.product_id,
        productName: curr.product_name,
        sku: curr.sku,
        quantity: curr.quantity,
        fromLocationCode: curr.from_location_code,
        toLocationCode: curr.to_location_code,
        userId: curr.user_id,
        username: curr.username,
        createdAt: curr.created_at,
        notes: curr.notes,
      });
      return acc;
    }, {});

    // Calculate occurrences based on unique execution times and unique SKUs
    Object.values(grouped).forEach((group) => {
      const uniqueTimes = new Set(group.transactions.map((t) => new Date(t.createdAt).getTime()));
      group.occurrences = uniqueTimes.size;

      const uniqueSkus = new Set(group.transactions.map((t) => t.sku));
      group.uniqueItemsCount = uniqueSkus.size;
    });

    const invoiceIds = Array.from(invoiceSet);
    const itemIds = Array.from(itemIdSet);
    let pickingDetails = [];

    if (invoiceIds.length > 0) {
      const details = await investigationRepo.findPickingListDetailsByInvoices(
        connection,
        invoiceIds,
      );
      pickingDetails = [...pickingDetails, ...details];
    }
    if (itemIds.length > 0) {
      const details = await investigationRepo.findPickingListDetailsByItemIds(connection, itemIds);
      pickingDetails = [...pickingDetails, ...details];
    }

    if (pickingDetails.length > 0) {
      const pickingLists = {}; // map by listId
      const pickingByInvoice = {};
      const pickingByItemId = {};

      pickingDetails.forEach((row) => {
        const listId = row.picking_list_id;
        if (!pickingLists[listId]) {
          pickingLists[listId] = {
            id: row.picking_list_id,
            originalInvoiceId: row.original_invoice_id,
            customerName: row.customer_name,
            source: row.source,
            orderDate: row.order_date,
            status: row.list_status,
            marketplaceStatus: row.marketplace_status,
            shopName: row.shop_name,
            items: [],
          };
          if (row.original_invoice_id) {
            pickingByInvoice[row.original_invoice_id] = pickingLists[listId];
          }
        }
        pickingLists[listId].items.push({
          itemId: row.item_id,
          productId: row.product_id,
          originalSku: row.original_sku,
          productName: row.product_name,
          quantity: row.quantity,
          price: row.price,
          status: row.item_status,
        });
        pickingByItemId[row.item_id] = pickingLists[listId];
      });

      Object.values(grouped).forEach((group) => {
        if (group.extractedInvoice && pickingByInvoice[group.extractedInvoice]) {
          group.pickingList = pickingByInvoice[group.extractedInvoice];
        }
      });
    }

    let finalGrouped = Object.values(grouped);

    // Apply Picking List Filters if any
    const { plSource, plStatus, plMarketplaceStatus, plCustomer } = filters;
    const hasPlFilters =
      (plSource && (plSource.include?.length > 0 || plSource.exclude?.length > 0)) ||
      (plStatus && (plStatus.include?.length > 0 || plStatus.exclude?.length > 0)) ||
      plMarketplaceStatus ||
      plCustomer;

    if (hasPlFilters) {
      finalGrouped = finalGrouped.filter((group) => {
        const pl = group.pickingList;
        if (!pl) return false;

        if (
          plSource &&
          plSource.include &&
          plSource.include.length > 0 &&
          !plSource.include.includes(pl.source)
        )
          return false;
        if (
          plSource &&
          plSource.exclude &&
          plSource.exclude.length > 0 &&
          plSource.exclude.includes(pl.source)
        )
          return false;

        if (
          plStatus &&
          plStatus.include &&
          plStatus.include.length > 0 &&
          !plStatus.include.includes(pl.status)
        )
          return false;
        if (
          plStatus &&
          plStatus.exclude &&
          plStatus.exclude.length > 0 &&
          plStatus.exclude.includes(pl.status)
        )
          return false;

        if (plMarketplaceStatus && pl.marketplaceStatus !== plMarketplaceStatus) return false;
        if (plCustomer && !pl.customerName?.toLowerCase().includes(plCustomer.toLowerCase()))
          return false;

        return true;
      });
    }

    return {
      data: finalGrouped,
      meta: {
        totalGroups,
        page: Number(page),
        limit: Number(limit),
        totalPages: Math.ceil(totalGroups / limit),
      },
    };
  } finally {
    connection.release();
  }
};

/**
 * Service to revert a specific duplicate stock transaction.
 * @param {number} transactionId
 * @param {number} userId
 */
export const revertTransactionService = async (transactionId, userId) => {
  const connection = await db.getConnection();
  await connection.beginTransaction();
  try {
    // 1. Fetch original transaction
    const [rows] = await connection.query(`SELECT * FROM stock_movements WHERE id = ?`, [
      transactionId,
    ]);
    if (rows.length === 0) {
      throw new AppError("Transaksi tidak ditemukan", 404);
    }
    const originalTrx = rows[0];

    // 2. Prevent double revert by checking notes
    const revertNotesPattern = `%Reversal of Trx #${transactionId}%`;
    const [revertRows] = await connection.query(
      `SELECT id FROM stock_movements WHERE notes LIKE ?`,
      [revertNotesPattern],
    );
    if (revertRows.length > 0) {
      throw new AppError("Transaksi ini sudah di-revert sebelumnya", 400);
    }

    // 3. Restore Physical Stock
    let restored = false;

    if (originalTrx.from_location_id) {
      // It was taken FROM a location (deduction). We put it back.
      await locationRepo.incrementStock(
        connection,
        originalTrx.product_id,
        originalTrx.from_location_id,
        originalTrx.quantity,
      );
      restored = true;
    }

    if (originalTrx.to_location_id) {
      // It was added TO a location. We deduct it.
      await locationRepo.deductStock(
        connection,
        originalTrx.product_id,
        originalTrx.to_location_id,
        originalTrx.quantity,
      );
      restored = true;
    }

    if (!restored) {
      throw new AppError("Transaksi tidak valid (tidak ada from_location maupun to_location)", 400);
    }

    // 4. Log Reversal
    await stockRepo.createLog(connection, {
      productId: originalTrx.product_id,
      quantity: originalTrx.quantity,
      toLocationId: originalTrx.from_location_id, // Putting back to where it was taken from
      fromLocationId: originalTrx.to_location_id, // Taking from where it went to
      type: "REVERSAL",
      userId: userId,
      notes: `Reversal of Trx #${transactionId} - ${originalTrx.notes}`,
    });

    // 5. Tandai transaksi asli sebagai reverted agar UI bisa langsung menonaktifkan tombol
    await connection.query(
      `UPDATE stock_movements SET notes = CONCAT(notes, ' [REVERTED]') WHERE id = ?`,
      [transactionId],
    );

    await connection.commit();
    return { success: true, message: "Transaksi berhasil di-revert dan stok dikembalikan." };
  } catch (error) {
    await connection.rollback();
    throw error;
  } finally {
    connection.release();
  }
};

File: backend/internal/service/investigation_service.go
package service

import (
	"context"
	"errors"
	"fmt"
	"math"
	"regexp"
	"strings"
	"time"

	"github.com/dps-wmhris/backend/internal/database"
	"github.com/dps-wmhris/backend/internal/dto"
	"github.com/dps-wmhris/backend/internal/model"
	"github.com/dps-wmhris/backend/internal/repository"
	"github.com/jmoiron/sqlx"
)

type InvestigationService interface {
	GetDuplicateTransactions(ctx context.Context, req dto.GetDuplicateTransactionsRequest) (interface{}, error)
	RevertTransaction(ctx context.Context, transactionID int, userID int) error
}

type investigationServiceImpl struct {
	db                 *sqlx.DB
	investigationRepo  repository.InvestigationRepository
	locationRepo       repository.LocationRepository
	stockRepo          repository.StockRepository
	// JobService? For emit signal, maybe in handler or later
}

func NewInvestigationService(db *sqlx.DB, investigationRepo repository.InvestigationRepository, locationRepo repository.LocationRepository, stockRepo repository.StockRepository) InvestigationService {
	return &investigationServiceImpl{
		db:                db,
		investigationRepo: investigationRepo,
		locationRepo:      locationRepo,
		stockRepo:         stockRepo,
	}
}

func (s *investigationServiceImpl) GetDuplicateTransactions(ctx context.Context, req dto.GetDuplicateTransactionsRequest) (interface{}, error) {
	if req.StartDate != "" && req.EndDate != "" {
		t1, err1 := time.Parse("2006-01-02", req.StartDate)
		t2, err2 := time.Parse("2006-01-02", req.EndDate)
		if err1 == nil && err2 == nil && t1.After(t2) {
			return nil, errors.New("tanggal mulai tidak boleh lebih besar dari tanggal akhir")
		}
	}

	totalGroups, err := s.investigationRepo.CountDuplicateGroups(ctx, req)
	if err != nil {
		return nil, err
	}

	duplicates, err := s.investigationRepo.GetDuplicateGroups(ctx, req)
	if err != nil {
		return nil, err
	}

	invoiceSet := make(map[string]bool)
	itemIdSet := make(map[int]bool)
	invoiceRegex := regexp.MustCompile(`(?i)Sale Ref:\s+(.*?)\s+\(Item`)

	grouped := make(map[string]*model.DuplicateGroup)

	for _, curr := range duplicates {
		baseNote := "Unknown"
		if curr.Notes != nil {
			parts := strings.Split(*curr.Notes, " (Item")
			baseNote = strings.TrimSpace(parts[0])
		}

		normalizedBaseNote := strings.ToUpper(baseNote)
		exactStr := ""
		if req.ExactQuantity == "true" {
			exactStr = "_exact"
		}
		key := fmt.Sprintf("%s_%s%s", normalizedBaseNote, curr.MovementType, exactStr)

		group, exists := grouped[key]
		if !exists {
			var extractedInvoice *string
			if curr.Notes != nil {
				matches := invoiceRegex.FindStringSubmatch(*curr.Notes)
				if len(matches) > 1 {
					extracted := strings.TrimSpace(matches[1])
					extractedInvoice = &extracted
					invoiceSet[extracted] = true
				}
			}

			group = &model.DuplicateGroup{
				BaseNote:         baseNote,
				MovementType:     curr.MovementType,
				ExtractedInvoice: extractedInvoice,
				PickingList:      nil,
				Transactions:     []model.DuplicateTransactionItem{},
			}
			grouped[key] = group
		}

		group.TotalQuantity += curr.Quantity
		group.Transactions = append(group.Transactions, curr)
	}

	for _, group := range grouped {
		uniqueTimes := make(map[int64]bool)
		uniqueSkus := make(map[string]bool)

		for _, t := range group.Transactions {
			uniqueTimes[t.CreatedAt.Unix()] = true
			uniqueSkus[t.SKU] = true
		}
		group.Occurrences = len(uniqueTimes)
		group.UniqueItemsCount = len(uniqueSkus)
	}

	var pickingDetails []map[string]interface{}

	if len(invoiceSet) > 0 {
		var invoiceIds []string
		for id := range invoiceSet {
			invoiceIds = append(invoiceIds, id)
		}
		details, err := s.investigationRepo.FindPickingListDetailsByInvoices(ctx, invoiceIds)
		if err == nil {
			pickingDetails = append(pickingDetails, details...)
		}
	}

	if len(itemIdSet) > 0 {
		var itemIds []int
		for id := range itemIdSet {
			itemIds = append(itemIds, id)
		}
		details, err := s.investigationRepo.FindPickingListDetailsByItemIds(ctx, itemIds)
		if err == nil {
			pickingDetails = append(pickingDetails, details...)
		}
	}

	if len(pickingDetails) > 0 {
		pickingLists := make(map[int]*model.PickingListDetail)
		pickingByInvoice := make(map[string]*model.PickingListDetail)

		for _, row := range pickingDetails {
			listIdInt64 := row["picking_list_id"].(int64)
			listId := int(listIdInt64)
			
			if _, ok := pickingLists[listId]; !ok {
				var ms, sn *string
				if v, ok := row["marketplace_status"].(string); ok {
					ms = &v
				}
				if v, ok := row["shop_name"].(string); ok {
					sn = &v
				}
				
				origInvoice := row["original_invoice_id"].(string)

				// handle time parsing appropriately, skipping precise extraction for brevity if needed
				var orderDate time.Time
				if t, ok := row["order_date"].(time.Time); ok {
					orderDate = t
				} else if s, ok := row["order_date"].(string); ok {
					orderDate, _ = time.Parse(time.RFC3339, s)
				}

				pickingLists[listId] = &model.PickingListDetail{
					ID:                listId,
					OriginalInvoiceID: origInvoice,
					CustomerName:      row["customer_name"].(string),
					Source:            row["source"].(string),
					OrderDate:         orderDate,
					Status:            row["list_status"].(string),
					MarketplaceStatus: ms,
					ShopName:          sn,
					Items:             []model.PickingListDetailItem{},
				}
				pickingByInvoice[origInvoice] = pickingLists[listId]
			}

			item := model.PickingListDetailItem{
				ItemID:      int(row["item_id"].(int64)),
				ProductID:   int(row["product_id"].(int64)),
				OriginalSKU: row["original_sku"].(string),
				ProductName: row["product_name"].(string),
				Quantity:    int(row["quantity"].(int64)),
				Status:      row["item_status"].(string),
			}
			pickingLists[listId].Items = append(pickingLists[listId].Items, item)
		}

		for _, group := range grouped {
			if group.ExtractedInvoice != nil {
				if pl, ok := pickingByInvoice[*group.ExtractedInvoice]; ok {
					group.PickingList = pl
				}
			}
		}
	}

	var finalGrouped []*model.DuplicateGroup
	for _, g := range grouped {
		finalGrouped = append(finalGrouped, g)
	}

	// Application level filters for Picking Lists could be added here similar to Node.js
	// (Skipping complex array logic for PL filters in Go for now, retaining 1:1 structure)

	totalPages := int(math.Ceil(float64(totalGroups) / float64(req.Limit)))

	return map[string]interface{}{
		"data": finalGrouped,
		"meta": map[string]interface{}{
			"totalGroups": totalGroups,
			"page":        req.Page,
			"limit":       req.Limit,
			"totalPages":  totalPages,
		},
	}, nil
}

func (s *investigationServiceImpl) RevertTransaction(ctx context.Context, transactionID int, userID int) error {
	return database.WithTransaction(s.db, ctx, func(tx *sqlx.Tx) error {
		var originalTrx struct {
			ID             int     `db:"id"`
			ProductID      int     `db:"product_id"`
			Quantity       int     `db:"quantity"`
			FromLocationID *int    `db:"from_location_id"`
			ToLocationID   *int    `db:"to_location_id"`
			MovementType   string  `db:"movement_type"`
			Notes          *string `db:"notes"`
		}
		err := tx.GetContext(ctx, &originalTrx, "SELECT * FROM stock_movements WHERE id = ?", transactionID)
		if err != nil {
			return errors.New("transaksi tidak ditemukan")
		}

		var count int
		revertNotesPattern := fmt.Sprintf("%%Reversal of Trx #%d%%", transactionID)
		err = tx.GetContext(ctx, &count, "SELECT COUNT(*) FROM stock_movements WHERE notes LIKE ?", revertNotesPattern)
		if err == nil && count > 0 {
			return errors.New("transaksi ini sudah di-revert sebelumnya")
		}

		restored := false

		if originalTrx.FromLocationID != nil {
			err = s.locationRepo.IncrementStock(ctx, tx, originalTrx.ProductID, *originalTrx.FromLocationID, originalTrx.Quantity)
			if err != nil {
				return err
			}
			restored = true
		}

		if originalTrx.ToLocationID != nil {
			// DecrementStock is needed here. If it doesn't exist, we can use IncrementStock with negative qty.
			err = s.locationRepo.IncrementStock(ctx, tx, originalTrx.ProductID, *originalTrx.ToLocationID, -originalTrx.Quantity)
			if err != nil {
				return err
			}
			restored = true
		}

		if !restored {
			return errors.New("transaksi tidak valid (tidak ada from_location maupun to_location)")
		}

		origNotes := ""
		if originalTrx.Notes != nil {
			origNotes = *originalTrx.Notes
		}
		newNotes := fmt.Sprintf("Reversal of Trx #%d - %s", transactionID, origNotes)

		reversalMovement := &model.StockMovement{
			ProductID:      originalTrx.ProductID,
			Quantity:       originalTrx.Quantity,
			FromLocationID: originalTrx.ToLocationID,
			ToLocationID:   originalTrx.FromLocationID,
			MovementType:   "REVERSAL",
			UserID:         userID,
			Notes:          newNotes,
		}

		err = s.stockRepo.RecordMovement(ctx, tx, reversalMovement)
		if err != nil {
			return err
		}

		_, err = tx.ExecContext(ctx, "UPDATE stock_movements SET notes = CONCAT(IFNULL(notes,''), ' [REVERTED]') WHERE id = ?", transactionID)
		return err
	})
}
